src/jvmMain/kotlin/io/beatmaps/api/review.kt

```diff
@@ -386,4 +429,145 @@ fun Route.reviewRoute() {
             }
         }
     }
+
+    post<ReplyApi.Create> { req ->
+        requireAuthorization { _, user ->
+            if (user.suspended) return@requireAuthorization call.respond(ActionResponse(false, listOf("Suspended account")))
+
+            val reply = call.receive<ReplyRequest>()
+
+            val created = newSuspendedTransaction {
+                val allowedUsers = Review
+                    .join(Beatmap, JoinType.LEFT, Review.mapId, Beatmap.id)
+                    .select(Review.userId, Beatmap.uploader)
+                    .where { Review.id eq req.reviewId }
+                    .firstOrNull()
+                    ?.let {
+                        listOf(it[Review.userId].value, it[Beatmap.uploader].value)
+                    } ?: listOf()
+
+                if (user.userId !in allowedUsers) {
+                    call.respond(ActionResponse(false, listOf("Unauthorised")))
+                    return@newSuspendedTransaction false
+                }
+
+                ReviewReply.insert {
+                    it[userId] = user.userId
+                    it[reviewId] = req.reviewId
+                    it[text] = reply.text
+                    it[createdAt] = NowExpression(createdAt)
+                    it[updatedAt] = NowExpression(updatedAt)
+                }.insertedCount > 0
+            }
+
+            if (created) {
```

Comment:
The way you suggest is how I had written it previously. I changed it to what I have now to prevent responding to the call twice and erasing the error message in the process, as in some cases an error does get returned when something goes wrong. If that's not something that can happen, I'll change it back to what I had.

Reply:
Oh I see, it's not clear how it's written that the alternative is the request has already got a response.
Is there any case where insertedCount would actually be 0?

`put<ReviewApi.Single>` at least ends with true which means you're guaranteed to use that response if you reach that code

Reply:
Actually should we just return an ActionResponse from the transaction?
Then just call.respond it after

Reply:
Could quite easily do the same for the other call if we're not bothered about the forbidden status code:

```kt
             captchaIfPresent(update.captcha) {
-                val success = newSuspendedTransaction {
+                val response = newSuspendedTransaction {
                     if (single.userId != sess.userId && !sess.isCurator()) {
-                        call.respond(HttpStatusCode.Forbidden)
-                        return@newSuspendedTransaction false
+                        return@newSuspendedTransaction ActionResponse(false, listOf("Forbidden"))
                     }

                     if (sess.suspended || UserDao[sess.userId].suspendedAt != null) {
                         // User is suspended
-                        call.respond(ActionResponse(false, listOf("Suspended account")))
-                        return@newSuspendedTransaction false
+                        return@newSuspendedTransaction ActionResponse(false, listOf("Suspended account"))
                     }

                     val oldData = if (single.userId != sess.userId) {
@@ -256,8 +253,7 @@ fun Route.reviewRoute() {

                         if (map.uploaderId.value == single.userId) {
                             // Can't review your own map
-                            call.respond(ActionResponse(false, listOf("Own map")))
-                            return@newSuspendedTransaction false
+                            return@newSuspendedTransaction ActionResponse(false, listOf("Own map"))
                         }

                         val isCollaborator = map.collaborators.values.any { singleCollaborator ->
@@ -266,8 +262,7 @@ fun Route.reviewRoute() {

                         if (isCollaborator) {
                             // Can't review maps that you collaborated on
-                            call.respond(ActionResponse(false, listOf("You're a collaborator of this map")))
-                            return@newSuspendedTransaction false
+                            return@newSuspendedTransaction ActionResponse(false, listOf("You're a collaborator of this map"))
                         }

                         Review.upsert(conflictIndex = Index(listOf(Review.mapId, Review.userId), true, "review_unique")) { r ->
@@ -301,14 +296,14 @@ fun Route.reviewRoute() {
                         )
                     }

-                    true
+                    ActionResponse(true, listOf())
                 }

-                if (success) {
+                if (response.success) {
                     val updateType = if (update.captcha == null) "updated" else "created"
                     call.pub("beatmaps", "reviews.$updateMapId.$updateType", null, ReviewUpdateInfo(updateMapId, single.userId))
-                    call.respond(ActionResponse(true, listOf()))
                 }
+                call.respond(response)
             }
         }
     }
```

Reply:
Sure, I don't mind that solution.